File: translated/f-to-c/DRB066-pointernoaliasing-orig-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Freshly allocated pointers do not alias to each other. No data race pairs.

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

void setup(int N)
{
    int i;
    double* m_pdv_sum;
    double* m_nvol;
    double* tar1;
    double* tar2;

    m_pdv_sum = (double*)malloc(N * sizeof(double));
    m_nvol = (double*)malloc(N * sizeof(double));
    tar1 = (double*)malloc(N * sizeof(double));
    tar2 = (double*)malloc(N * sizeof(double));

    m_pdv_sum = tar1;
    m_nvol = tar2;

    #pragma omp parallel for schedule(static)
    for (i = 0; i < N; i++) {
        tar1[i] = 0.0;
        tar2[i] = (i + 1) * 2.5;  // Adjust for 0-based indexing
    }

    // printf("%f %f\n", tar1[N-1], tar2[N-1]);
    
    // In C, we just need to free the allocated memory
    free(tar1);
    free(tar2);
    // Note: m_pdv_sum and m_nvol were pointing to tar1 and tar2, 
    // so we don't free them separately
}

int main()
{
    int N = 1000;
    setup(N);
    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB066-pointernoaliasing-orig-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Freshly allocated pointers do not alias to each other. No data race pairs.

package main

import (
	"runtime"
	"sync"
)

// Module DRB066 translated to package-level functions
func setup(N int) {
	var tar1, tar2 []float64

	tar1 = make([]float64, N)
	tar2 = make([]float64, N)

	//$omp parallel do schedule(static)
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := N / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= N; start += chunkSize {
		end := start + chunkSize - 1
		if end > N {
			end = N
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				tar1[i-1] = 0.0
				tar2[i-1] = float64(i) * 2.5
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end parallel do

	// Uncomment to print values for checking
	// fmt.Printf("%f %f\n", tar1[N-1], tar2[N-1])
}

func main() {
	var N int
	N = 1000

	setup(N)
}
